x-pack/legacy/plugins/lens/public/indexpattern_plugin/dimension_panel/dimension_panel.tsx

```diff
@@ -133,24 +151,38 @@ export const IndexPatternDimensionPanel = memo(function IndexPatternDimensionPan
           const newColumn = hasFieldChanged
             ? changeField(selectedColumn, currentIndexPattern, droppedItem.field)
             : buildColumn({
-                columns: props.state.layers[props.layerId].columns,
+                columns: layer.columns,
                 indexPattern: currentIndexPattern,
-                layerId,
+                layerId: props.layerId,
                 suggestedPriority: props.suggestedPriority,
                 field: droppedItem.field,
               });
 
-          props.setState(
-            changeColumn({
-              state: props.state,
-              layerId,
+          if (droppedItem.movedColumn && props.onRemove) {
+            props.onRemove(droppedItem.movedColumn);
+          }
+
+          props.setState(currentState => {
+            const withChangedColumn = changeColumn({
+              state: currentState,
+              layerId: props.layerId,
               columnId: props.columnId,
               newColumn,
               // If the field has changed, the onFieldChange method needs to take care of everything including moving
               // over params. If we create a new column above we want changeColumn to move over params.
               keepParams: !hasFieldChanged,
-            })
-          );
+            });
+
+            if (!droppedItem.movedColumn) {
+              return withChangedColumn;
+            }
+
+            return deleteColumn({
```

Comment:
Should we remove or switch? I'm not sure but I was slightly surprised when trying it out for the first time.

Reply:
Yeah. My thought was to swap them, too, but the card in our backlog said that this should be a move / replace operation. I can see the argument for either, and it shouldn't be too hard to change if we want.